# Analyzing data crawling from eBay Kleinanzeigen

in this project, we aim to cleaning dataset from eBay Kleinanzeigen and analyzing it.

eBay Kleinanzeigen dataset, is data crawling from eBay Website with table, contain information about Car selling, this sample has 50.000 data point

## import dataset

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

after that, using function info to get detail information about dataset, such as number non null rows, type column and others

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

we can analyze that there was 20 column with 15 object columns or non-numeric columns and 5 numeric columns, in that info, we know that 5 object columns have null value, with conclusion that, we must handling object type column with null value

In [3]:
#using function head to get five table of dataset
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

using head() function, we know some information

1. in the first column there was row with datetime, but read by object

2. second row have mixed upper character and lower character

3. column price has value number, but read as object because $ in that column

4. same as column price, odoo column have value number , but read as object

5. in column brand, especially in the name column, have spaces, that should be remove

6. datecreated has same value as first column and last seen column


## change column name

in this section, i will change the column name, to make data easy to analysis

In [4]:
#create new list
new_column = []

#loop the data columns
for col in autos.columns:
    col = col.replace('yearOfRegistration', 'registration_year')
    col = col.replace('monthOfRegistration', 'registration_month')
    col = col.replace('notRepairedDamage', 'unrepaired_damage')
    col = col.replace('dateCreated', 'ad_created')
    col = col.strip()
    col = col.lower()
    new_column.append(col)
    
#replace columns with the list of new columns
autos.columns = new_column

#check the data 
autos.head()


datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

for this part, i change many name of column, such as year of registrion, there are many reason why i should change the name:

1. there are many column have mixed upper and lower case, so to make easy when we analyze the data, it will be good, to change all of that column name become lower case.

2. there are one column have space or snakecase, we should change it and remove spaces between column's name

3. we also change many name of column such as yearof registration to registration year, just to make easy when we want analyse more deeper in that columns

## data exploration

In [5]:
#use describe to get detail information about columns
autos.describe(include = 'all')

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

i got many information, such as:

1. datecrawled have 48213 unique data and 50000 row, it's mean that there was many duplicat data
2. same as datecrawled, name column have duplicate data, need more deply analysis
3. we should change column price, with remove $ and change type to integer
4. we should more concern in registration, because min, and max have 'out of the box' number. for instance,  it's imposible car registered at year 1000 because, at that time, there was no car. second, 9999 years also impossible number beacause, we take the data at 2016.
5. focus on vehicle type, we must handle missing value. also for gearbox, model and fueltype column
6. we should more concern with powerapps column, beacuse they have outliers 
7. we should change type data in odometer column, become integer
8. last seen columns have many duplicat datetime data
9. i think we should remove no of pictures column, because all of row have zero number, and it don't give any clue for analysis
10. we should change language, from germany to english to make easy for further analysis

In [6]:
#replace and change type of price and odometer
autos['price'] = autos['price'].str.replace("$",'').str.replace(",","").astype('int')
autos['odometer'] = autos['odometer'].str.replace("km","").str.replace(",","").astype('int')

In [7]:
#rename dataframe
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

In [8]:
#check the data
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype  price   abtest vehicletype  registration_year    gearbox  \
0  privat   Angebot   5000  control         bus               2004    manuell   
1  privat   Angebot   8500  control   limousine               1997  automatik   
2  privat   Angebot   8990     test   limousine               2009    manuell   
3  privat   Angebot   4350  control  kleinwagen               2007  automatik   
4  privat   Angebot   1350     test       kombi               2003    manuell   

   powerps   model  odometer_km  registration_month fueltype       brand  \
0      158  andere       150000                   3      lpg     peugeot   
1      286     7er       150000                   6   benzin         bmw   
2      102    golf        70000                   7   benzin  volkswagen   
3       71  fortwo        70000                   6   benzin       smart   
4        0   focus       150000                   7   benzin        ford   

  unrepaired_damage           ad_created  nrofpictures  postalcode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   
3              nein  2016-03-12 00:00:00             0       33729   
4              nein  2016-04-01 00:00:00             0       39218   

              lastseen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [9]:
#analyst odometer and price
print(autos['odometer_km'].unique().shape)
print(autos['price'].unique().shape)

(13,)
(2357,)


using unique, we know that most of data have same odometer, lets we looks using describe

In [10]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [11]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

when we using desribe, there was some interesting information from that
1. in odometer column 50 %, 75% and max column have same number, so i think that most of odometer have value in 150000
2. there was outlier in the price column, we can know from max and min

for further information, let's take a look deply using value counts

In [12]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

most of data has value 15000, and i think there was no outlier in there, so we will focus on price column, which is have many outlier

In [13]:
autos['price'].value_counts().sort_index(ascending=False).head()

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price, dtype: int64

using sort index, ascending false, there was one outlier, with value 9999999

In [14]:
autos['price'].value_counts().sort_index(ascending=True).head(30)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
Name: price, dtype: int64

from my point of view, i think impossible car in ebay have 0 dollar, so i will categorize it with outliers and i will remove it

In [15]:
#remove the rows with outliers
autos = autos[(autos['price'] > 0) & (autos['price'] < 99999999)]

In [16]:
#check the data shape
autos.shape

(48578, 20)

## analyzing data datetime

in this section, we will explore the column with datetime value

In [17]:
#analyze the distribution of date_crawled, ad_created and last_seen dataset

autos['datecrawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()



2016-03-05    0.025320
2016-03-06    0.014039
2016-03-07    0.036004
2016-03-08    0.033328
2016-03-09    0.033101
2016-03-10    0.032175
2016-03-11    0.032566
2016-03-12    0.036930
2016-03-13    0.015666
2016-03-14    0.036539
2016-03-15    0.034275
2016-03-16    0.029602
2016-03-17    0.031640
2016-03-18    0.012907
2016-03-19    0.034769
2016-03-20    0.037877
2016-03-21    0.037404
2016-03-22    0.032998
2016-03-23    0.032216
2016-03-24    0.029334
2016-03-25    0.031599
2016-03-26    0.032196
2016-03-27    0.031084
2016-03-28    0.034851
2016-03-29    0.034131
2016-03-30    0.033678
2016-03-31    0.031846
2016-04-01    0.033678
2016-04-02    0.035469
2016-04-03    0.038598
2016-04-04    0.036519
2016-04-05    0.013092
2016-04-06    0.003170
2016-04-07    0.001400
Name: datecrawled, dtype: float64

from date crawled , we know that most of data crawled at april and march 2016 

In [18]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033163
2016-03-10    0.031887
2016-03-11    0.032896
2016-03-12    0.036766
2016-03-13    0.017004
2016-03-14    0.035181
2016-03-15    0.034007
2016-03-16    0.030117
2016-03-17    0.031290
2016-03-18    0.013586
2016-03-19    0.033678
2016-03-20    0.037939
2016-03-21 

same as datcrawled, most of ad_created dataset come form march and april, it can be seen in below:

In [19]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False, ascending=False)


2016-04-03    0.038845
2016-03-20    0.037939
2016-03-21    0.037610
2016-04-04    0.036889
2016-03-12    0.036766
2016-03-14    0.035181
2016-04-02    0.035139
2016-03-28    0.034975
2016-03-07    0.034728
2016-03-29    0.034069
2016-03-15    0.034007
2016-04-01    0.033678
2016-03-19    0.033678
2016-03-30    0.033493
2016-03-08    0.033348
2016-03-09    0.033163
2016-03-11    0.032896
2016-03-22    0.032813
2016-03-26    0.032257
2016-03-23    0.032052
2016-03-10    0.031887
2016-03-31    0.031887
2016-03-25    0.031743
2016-03-17    0.031290
2016-03-27    0.030981
2016-03-16    0.030117
2016-03-24    0.029273
2016-03-05    0.022891
2016-03-13    0.017004
2016-03-06    0.015316
                ...   
2016-02-20    0.000041
2016-02-18    0.000041
2016-02-26    0.000041
2016-02-14    0.000041
2016-02-02    0.000041
2016-02-12    0.000041
2016-02-05    0.000041
2016-01-10    0.000041
2016-02-24    0.000041
2016-02-07    0.000021
2016-01-22    0.000021
2015-12-05    0.000021
2015-12-30 

In [20]:
autos['datecrawled'].str[:10].value_counts(normalize=True, dropna=False, ascending=False)


2016-04-03    0.038598
2016-03-20    0.037877
2016-03-21    0.037404
2016-03-12    0.036930
2016-03-14    0.036539
2016-04-04    0.036519
2016-03-07    0.036004
2016-04-02    0.035469
2016-03-28    0.034851
2016-03-19    0.034769
2016-03-15    0.034275
2016-03-29    0.034131
2016-04-01    0.033678
2016-03-30    0.033678
2016-03-08    0.033328
2016-03-09    0.033101
2016-03-22    0.032998
2016-03-11    0.032566
2016-03-23    0.032216
2016-03-26    0.032196
2016-03-10    0.032175
2016-03-31    0.031846
2016-03-17    0.031640
2016-03-25    0.031599
2016-03-27    0.031084
2016-03-16    0.029602
2016-03-24    0.029334
2016-03-05    0.025320
2016-03-13    0.015666
2016-03-06    0.014039
2016-04-05    0.013092
2016-03-18    0.012907
2016-04-06    0.003170
2016-04-07    0.001400
Name: datecrawled, dtype: float64

but there was many different:
1. in data crawled dataset, most of data come from march and april, when ad_created from january to april, it's mean that data crawled from website in march and april when ads from website was been created in january

last, from last_seen dataset

In [21]:
autos['lastseen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()


2016-03-05    0.001070
2016-03-06    0.004323
2016-03-07    0.005393
2016-03-08    0.007431
2016-03-09    0.009613
2016-03-10    0.010663
2016-03-11    0.012372
2016-03-12    0.023797
2016-03-13    0.008893
2016-03-14    0.012598
2016-03-15    0.015871
2016-03-16    0.016448
2016-03-17    0.028079
2016-03-18    0.007349
2016-03-19    0.015830
2016-03-20    0.020647
2016-03-21    0.020647
2016-03-22    0.021368
2016-03-23    0.018527
2016-03-24    0.019762
2016-03-25    0.019206
2016-03-26    0.016798
2016-03-27    0.015645
2016-03-28    0.020874
2016-03-29    0.022356
2016-03-30    0.024764
2016-03-31    0.023797
2016-04-01    0.022788
2016-04-02    0.024929
2016-04-03    0.025197
2016-04-04    0.024476
2016-04-05    0.124768
2016-04-06    0.221808
2016-04-07    0.131912
Name: lastseen, dtype: float64

last seen and date_crawled have connection each other, one of that is data created from march and april, it's mean that datecrawled and lastseen dataset had been taken at same interval, let's check registration year

In [22]:
autos['registration_year'].describe()

count    48578.000000
mean      2004.753119
std         88.632571
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [23]:
autos['registration_year'].value_counts().sort_index()

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      24
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2637
2002    2486
2003    2699
2004    2703
2005    2936
2006    2671
2007    2277
2008    2215
2009    2086
2010    1589
2011    1623
2012    1310
2013     803
2014     664
2015     392
2016    1220
2017    1393
2018     471
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

from registration year, we found many outliers, such as year 1000, 1111, 1001, 8888, 9000 and 9999, so we will delete it from dataset

In [24]:
autos.loc[(autos['registration_year'] < 1900) | (autos['registration_year'] > 2016),:]

datecrawled                                               name  \
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
65     2016-04-04 19:30:39                      Ford_Fiesta_zum_ausschlachten   
68     2016-04-03 17:36:59  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   
84     2016-03-27 19:52:54                                     Renault_twingo   
113    2016-04-03 14:58:29                              Golf_4_Anfaenger_auto   
164    2016-03-13 20:39:16  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...   
197    2016-04-05 10:36:24                              VW_Polo_9N_an_Bastler   
253    2016-03-27 13:25:18             Ford_mondeo_Gas_anlage_mit_TÜV_04.2017   
348    2016-03-17 20:58:24  VW_Beetle_1.8Turbo_mit_Vollausstattung_und_seh...   
390    2016-03-25 12:59:06   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19   
438    2016-03-10 20:36:25                VW_Golf_5_V_1.6_2004_Klima_Tempomat   
453    2016-03-28 13:51:12                                         Armee_Jeep   
454    2016-03-25 09:57:03    Ford_Fiesta_Servo_D3_Kat_86€_Steuern_65.000_tkm   
457    2016-03-15 17:48:39  Mercedes_Benz_W203_C200_Kombi_Kompressor_c_kla...   
477    2016-03-31 08:53:34                                         Citroën_C4   
545    2016-03-22 18:50:53                                      Bmw_x5_/__E53   
548    2016-03-10 17:44:57            Peugeot_206_1.9_Hdi_4_Tueren_Tuev_02.17   
557    2016-03-16 03:36:34                           Smart_Fortwo_Passion_MHD   
568    2016-03-18 09:48:44             PRIMA_ANFÄNGERAUTO_smart_fortwo_cabrio   
577    2016-04-03 14:49:09                                  Volkswagen_Golf_3   
582    2016-03-31 11:57:46                                    V10_TDI_TOUAREG   
654    2016-04-01 15:48:34              Ich_verkaufe_meinen_gepflegten_daewoo   
740    2016-04-02 19:49:46                            Seat_Leon_zu_verkaufen!   
754    2016-03-26 13:47:56                         Ford_Ka_Kleinwagen_BJ_2008   
758    2016-03-12 18:42:33                            Honda_Civic_mit_TÜV_!!!   
765    2016-03-26 11:52:53               BMW_3er_Limousine_mit_1_Jahr_TÜV!!!!   
802    2016-03-19 11:37:23                 Lada_mit_wenig_km_neuem_Tuev_bj_08   
871    2016-04-04 19:38:38                                   Citroen_C2_60_PS   
889    2016-03-12 16:52:19                          Verkaufe_Opel_Corsa_C_1_2   
...                    ...                                                ...   
49130  2016-03-09 18:46:05                                      Renault_Modus   
49153  2016-03-12 01:36:59                                        Corsa_c20xe   
49154  2016-03-24 17:51:22                    Auto_mit_TÜV_bis_September_2017   
49178  2016-03-24 19:58:43       Mazda_323_F_TÜV__neu_2te_Hand_viele_Neuteile   
49185  2016-03-28 11:45:31              Alfa_Romeo_156_Sportwagon_1_8_T_Spark   
49218  2016-04-01 19:48:06          Ford_Fiesta_1.3_4/5_Tuerer_16_Monate_Tuev   
49245  2016-03-26 08:32:30                         Audi_A3_1.8T_tausch_moegl.   
49259  2016-03-19 18:45:05                  Volkswagen_Lupo_TÜV_06.2017_Klima   
49261  2016-03-10 12:51:40                                        Toyota_aygo   
49262  2016-03-26 18:31:21                                        Opel_vectra   
49266  2016-03-10 00:54:09  Toyota_Corolla_Verso_7_Sitzer_104.000km_2_2_Di...   
49283  2016-03-15 18:38:53                                         Citroen_HY   
49343  2016-03-12 11:56:55                                   bmw_320i_touring   
49347  2016-03-21 20:41:09                         Honda_civic_mit_TÜV_3/2017   
49354  2016-04-05 10:31:29                        Bmw_e39_523i_mit_neuem_Tuev   
49389  2016-03-10 07:54:10                                         Astra_G_CC   
49411  2016-03-28 12:58:19                            Renault_twingo_Tuev_neu   
49522  2016-03-14 10:42:08               Mazda

from my point of view  i will aceptable year 2017,2018, because  in that year, number of data more than 300, so if i delete it, it can change our analysis.

In [25]:
autos = autos.loc[(autos['registration_year'] > 1900) & (autos['registration_year'] < 2019),:]

In [26]:
autos['registration_year'].value_counts(normalize=True, ascending=False)

2000    0.064997
2005    0.060466
1999    0.059684
2004    0.055668
2003    0.055585
2006    0.055009
2001    0.054308
2002    0.051199
1998    0.048665
2007    0.046894
2008    0.045617
2009    0.042961
1997    0.040180
2011    0.033425
2010    0.032725
2017    0.028689
1996    0.028277
2012    0.026979
1995    0.025290
2016    0.025126
2013    0.016538
2014    0.013675
1994    0.012954
2018    0.009700
1993    0.008753
2015    0.008073
1992    0.007641
1990    0.007146
1991    0.006982
1989    0.003583
          ...   
1966    0.000453
1976    0.000432
1969    0.000391
1975    0.000371
1965    0.000350
1964    0.000247
1963    0.000165
1961    0.000124
1959    0.000124
1910    0.000103
1958    0.000082
1937    0.000082
1956    0.000082
1962    0.000082
1950    0.000062
1954    0.000041
1941    0.000041
1951    0.000041
1957    0.000041
1955    0.000041
1934    0.000041
1948    0.000021
1953    0.000021
1927    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.0000

most of data give information, that most of car which is sell in ebay,have registration year at 2000 or between 1998 - 2009, it's mean that people likely want to sell their old car, maybe for buy brand new car, or they have more useful car.

## Exploring Price by Brand

In [27]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212826
bmw               0.108637
opel              0.108637
mercedes_benz     0.095766
audi              0.085839
ford              0.069693
renault           0.047883
peugeot           0.029451
fiat              0.026011
seat              0.018927
skoda             0.016064
nissan            0.015261
mazda             0.015220
smart             0.014293
citroen           0.014128
toyota            0.012583
hyundai           0.009947
sonstige_autos    0.009700
volvo             0.009062
mini              0.008609
mitsubishi        0.008176
honda             0.007970
kia               0.007105
alfa_romeo        0.006590
porsche           0.005911
suzuki            0.005890
chevrolet         0.005664
chrysler          0.003481
dacia             0.002657
daihatsu          0.002513
jeep              0.002204
subaru            0.002101
land_rover        0.002039
saab              0.001627
daewoo            0.001565
jaguar            0.001524
trabant           0.001380
r

using value counts, we know that most of brand in ebay, from uerope and usa,

our objective now aggregate data brand with other column, for this case, we will aggregate data brand with price column

for this analysis i will only focus on brand with number of sell in ebay more than 5 %, i choose that, because it can represent most of people who sell the car

In [28]:
brand_5percent = ['volkswagen', 'opel','bmw', 'mercedes_benz','audi','ford']

brand_dictionary = {}

for b in brand_5percent:
    brand = autos[autos['brand'] == b]
    mean_brand = brand['price'].mean()
    brand_dictionary[b] = mean_brand
    
    

sorted_data = sorted(brand_dictionary.items(), key = lambda x : x[1], reverse = True)

sorted_data

[('audi', 9212.9306621881),
 ('mercedes_benz', 8526.623225806452),
 ('bmw', 8493.857251184834),
 ('ford', 7305.141843971631),
 ('volkswagen', 6600.382233404296),
 ('opel', 5281.321706161138)]

we found many information, from this analysis:
1. for brand with number of sell data more 5% from all_data: audi , mercedes benz and bmw have highest average number of price rather than 'ford', 'volkswagen' and 'opel'.

i think the reason behind that is because the quality of brand itself, we know that , especially, mercedes_benz and bmw are car premium with highest price, so it's normal when that brand have highest price in ebay

let's compare with mileage

In [29]:
autos.columns

Index(['datecrawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer_km', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nrofpictures', 'postalcode',
       'lastseen'],
      dtype='object')

In [30]:
mileage_dictionary ={}

#loop the data
for b in brand_5percent:
    brand_data = autos[autos['brand'] == b]
    mean_mileage = brand_data['odometer_km'].mean()
    mileage_dictionary[b] = mean_mileage
    

In [31]:
price_brand_series = pd.Series(brand_dictionary)
mileage_brand_series = pd.Series(mileage_dictionary)

price_and_mileage = pd.DataFrame({'price_avg':price_brand_series, 'mileage_avg':mileage_brand_series}).sort_values('mileage_avg', ascending=False)

In [32]:
price_and_mileage

mileage_avg    price_avg
bmw            132686.255924  8493.857251
mercedes_benz  130848.387097  8526.623226
audi           129492.562380  9212.930662
opel           129455.924171  5281.321706
volkswagen     128961.679892  6600.382233
ford           124327.718676  7305.141844

from above, we know that bmw, mercedes, and audi reach top 3 higher average mileage. when opel, volkswagen and ford  under that brand, it's look like same as price avg, just different sequence in price_avg, for instance, in mileage_avg, bmw is the first when in price_avg become the third, same as opel.

from that dataframe, we can conclude that bmw, mercedes and audi with higher price tag have  high mileage_avg, but we should know, if we split that dataframe become two class, with first class are bmw , mercedes and audi , and second class( opel, volkswagen, ford) we will know, it's different with price.

brand with lower mileage avg will be have higher price and vice versa.

this trend maybe have connection with car machine and time, when mileage_avg is higher, it's mean that time usage that car is higher and make machine and car not have good condition, it's causing  price decrease. price decrease also have connection with people, when people want to buy car with higher price it also follow by good car machine condition

## Challange -> Cleaning german language and datetime dataset 

In [33]:
#changing german language

#check the data
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype  price   abtest vehicletype  registration_year    gearbox  \
0  privat   Angebot   5000  control         bus               2004    manuell   
1  privat   Angebot   8500  control   limousine               1997  automatik   
2  privat   Angebot   8990     test   limousine               2009    manuell   
3  privat   Angebot   4350  control  kleinwagen               2007  automatik   
4  privat   Angebot   1350     test       kombi               2003    manuell   

   powerps   model  odometer_km  registration_month fueltype       brand  \
0      158  andere       150000                   3      lpg     peugeot   
1      286     7er       150000                   6   benzin         bmw   
2      102    golf        70000                   7   benzin  volkswagen   
3       71  fortwo        70000                   6   benzin       smart   
4        0   focus       150000                   7   benzin        ford   

  unrepaired_damage           ad_created  nrofpictures  postalcode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   
3              nein  2016-03-12 00:00:00             0       33729   
4              nein  2016-04-01 00:00:00             0       39218   

              lastseen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

many columns like seller offertype, vehicletype have german language, let's change it to english

In [34]:
#change name in  columns
autos['seller'] = autos['seller'].str.replace('privat', 'private').str.replace('gewerblich', 'commercial')
autos['offertype'] = autos['offertype'].str.replace('Angebot','Offer')
autos['vehicletype'] = autos['vehicletype'].str.replace('kleinwagen','small car').str.replace('kombi','wagon').str.replace('andere','other')
autos['gearbox'] = autos['gearbox'].str.replace('manuell','manual').str.replace('automatik','automatic')
autos['model'] = autos['model'].str.replace('andere','other')
autos['fueltype'] = autos['fueltype'].str.replace('benzin', 'petrol')
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace('nein','no').str.replace('ja','yes')


In [35]:
#check again
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    seller offertype  price   abtest vehicletype  registration_year  \
0  private     Offer   5000  control         bus               2004   
1  private     Offer   8500  control   limousine               1997   
2  private     Offer   8990     test   limousine               2009   
3  private     Offer   4350  control   small car               2007   
4  private     Offer   1350     test       wagon               2003   

     gearbox  powerps   model  odometer_km  registration_month fueltype  \
0     manual      158   other       150000                   3      lpg   
1  automatic      286     7er       150000                   6   petrol   
2     manual      102    golf        70000                   7   petrol   
3  automatic       71  fortwo        70000                   6   petrol   
4     manual        0   focus       150000                   7   petrol   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot                no  2016-03-26 00:00:00             0   
1         bmw                no  2016-04-04 00:00:00             0   
2  volkswagen                no  2016-03-26 00:00:00             0   
3       smart                no  2016-03-12 00:00:00             0   
4        ford                no  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

change the column dates

In [36]:
#cleaning the date
def cleaning_date(data):
    cleaning = data.str[:10]
    data = cleaning.astype(str).str.replace('-','').astype(int)
    return data

cleaning_datecrawled = cleaning_date(autos['datecrawled'])
cleaning_ad_created = cleaning_date(autos['ad_created'])
cleaning_lastseen = cleaning_date(autos['lastseen'])

In [37]:
#change the data
autos['datecrawled'] = cleaning_datecrawled
autos['ad_created'] = cleaning_ad_created
autos['lastseen'] = cleaning_lastseen

In [38]:
#check the data
autos.head()

datecrawled                                               name   seller  \
0     20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1     20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2     20160326                         Volkswagen_Golf_1.6_United  private   
3     20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4     20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   

  offertype  price   abtest vehicletype  registration_year    gearbox  \
0     Offer   5000  control         bus               2004     manual   
1     Offer   8500  control   limousine               1997  automatic   
2     Offer   8990     test   limousine               2009     manual   
3     Offer   4350  control   small car               2007  automatic   
4     Offer   1350     test       wagon               2003     manual   

   powerps   model  odometer_km  registration_month fueltype       brand  \
0      158   other       150000                   3      lpg     peugeot   
1      286     7er       150000                   6   petrol         bmw   
2      102    golf        70000                   7   petrol  volkswagen   
3       71  fortwo        70000                   6   petrol       smart   
4        0   focus       150000                   7   petrol        ford   

  unrepaired_damage  ad_created  nrofpictures  postalcode  lastseen  
0                no    20160326             0       79588  20160406  
1                no    20160404             0       71034  20160406  
2                no    20160326             0       35394  20160406  
3                no    20160312             0       33729  20160315  
4                no    20160401             0       39218  20160401

### analyst damage columns

In [39]:
damage_dictionary={}
for d in autos['unrepaired_damage'].unique():
    damage_data = autos[autos['unrepaired_damage']==d]
    damage_mean = damage_data['price'].mean()
    damage_dictionary[d] = damage_mean

In [40]:
damage_dictionary

{nan: nan, 'yes': 2434.6313432835823, 'no': 8580.11361283644}

In [41]:
differences = damage_dictionary['no'] - damage_dictionary['yes']

differences

6145.482269552857

from data unrepaired_damage we know, when car have unrepaired parts, that car become more cheaper than car with no damage, at least - if we using price average- have 6000 dollar different